diff --git a/src/path/temp.c b/src/path/temp.c
index bbcfb5f..583a3a7 100644
--- a/src/path/temp.c
+++ b/src/path/temp.c
@@ -10,6 +10,8 @@
 
 #include "cli/note.h"
 
+#define P_tmpdir (getenv("PROOT_TMPDIR") != NULL ? getenv("PROOT_TMPDIR") : "/data/local/tmp")
+
 /**
  * Remove recursively the content of the current working directory.
  * This latter has to lie in P_tmpdir (ie. "/tmp" on most systems).
