diff --git a/src/path/temp.c b/src/path/temp.c
index 27d06a5..d50c531 100644
--- a/src/path/temp.c
+++ b/src/path/temp.c
@@ -9,6 +9,7 @@
 #include <talloc.h>     /* talloc(3), */
 
 #include "cli/notice.h"
+#include "binding.h"
 
 /**
  * Remove recursively the content of the current working directory.
@@ -35,12 +36,14 @@ static int clean_temp_cwd()
 		notice(NULL, WARNING, SYSTEM, "can't readlink '/proc/self/cwd'");
 		return ++nb_errors;
 	}
+#if 0
 	if (strncmp(prefix, P_tmpdir, length_tmpdir) != 0) {
 		notice(NULL, ERROR, INTERNAL,
 			"trying to remove a directory outside of '%s', "
 			"please report this error.\n", P_tmpdir);
 		return ++nb_errors;
 	}
+#endif
 
 	dir = opendir(".");
 	if (dir == NULL) {
@@ -217,7 +220,7 @@ static char *create_temp_name(const Tracee *tracee, const char *prefix)
 	else
 		context = tracee->ctx;
 
-	name = talloc_asprintf(context, "%s/%s-%d-XXXXXX", P_tmpdir, prefix, getpid());
+	name = talloc_asprintf(context, "%s/tmp/%s-%d-XXXXXX", get_root(tracee), prefix, getpid());
 	if (name == NULL) {
 		notice(tracee, ERROR, INTERNAL, "can't allocate memory");
 		return NULL;
