diff --git a/src/path/temp.c b/src/path/temp.c
index bbcfb5f..b0d1b53 100644
--- a/src/path/temp.c
+++ b/src/path/temp.c
@@ -10,6 +10,17 @@
 
 #include "cli/note.h"
 
+static const char *proot_tmpdir = P_tmpdir;
+
+/* __attribute__((constructor)) is GCC-specific extension */
+__attribute__((constructor)) static void init_tmpdir()
+{
+	if (getenv("PROOT_TMPDIR") != NULL)
+		proot_tmpdir = getenv("PROOT_TMPDIR");
+}
+
+#define P_tmpdir proot_tmpdir
+
 /**
  * Remove recursively the content of the current working directory.
  * This latter has to lie in P_tmpdir (ie. "/tmp" on most systems).
